crates/utils/src/interval_map/mod.rs

```diff
@@ -0,0 +1,902 @@
+                        w = x.parent(NodeRef::right_owned);
+                    }
+                    w.set_color(x.parent(NodeRef::color));
+                    x.parent(NodeRef::set_color_fn(Color::Black));
+                    w.right(NodeRef::set_color_fn(Color::Black));
+                    x.parent(|p| self.left_rotate(p));
+                    x = self.root.clone_rc();
+                }
+            } else {
+                w = x.parent(NodeRef::left_owned);
+                if w.is_red() {
+                    w.set_color(Color::Black);
+                    x.parent(NodeRef::set_color_fn(Color::Red));
+                    x.parent(|p| self.right_rotate(p));
+                    w = x.parent(NodeRef::left_owned);
+                }
+                if w.is_sentinel() {
+                    break;
+                }
+                if w.left(NodeRef::is_black) && w.right(NodeRef::is_black) {
+                    w.set_color(Color::Red);
+                    x = x.parent_owned();
+                } else {
+                    if w.left(NodeRef::is_black) {
+                        w.right(NodeRef::set_color_fn(Color::Black));
+                        w.set_color(Color::Red);
+                        self.left_rotate(&w);
+                        w = x.parent(NodeRef::left_owned);
+                    }
+                    w.set_color(x.parent(NodeRef::color));
+                    x.parent(NodeRef::set_color_fn(Color::Black));
+                    w.left(NodeRef::set_color_fn(Color::Black));
+                    x.parent(|p| self.right_rotate(p));
+                    x = self.root.clone_rc();
+                }
+            }
+        }
+        x.set_color(Color::Black);
+    }
+
+    /// Binary tree left rotate.
+    fn left_rotate(&mut self, x: &NodeRef<T, V>) {
+        if x.right(NodeRef::is_sentinel) {
+            return;
+        }
+        let y = x.right_owned();
+        x.set_right(y.left_owned());
+        if !y.left(NodeRef::is_sentinel) {
+            y.left(|l| l.set_parent(x.clone_rc()));
+        }
+
+        self.replace_parent(x, &y);
+        y.set_left(x.clone_rc());
+
+        Self::rotate_update_max(x, &y);
+    }
+
+    /// Binary tree right rotate.
+    fn right_rotate(&mut self, x: &NodeRef<T, V>) {
+        if x.left(NodeRef::is_sentinel) {
+            return;
+        }
+        let y = x.left_owned();
+        x.set_left(y.right_owned());
+        if !y.right(NodeRef::is_sentinel) {
+            y.right(|r| r.set_parent(x.clone_rc()));
+        }
+
+        self.replace_parent(x, &y);
+        y.set_right(x.clone_rc());
+
+        Self::rotate_update_max(x, &y);
+    }
+
+    /// Replaces parent during a rotation.
+    fn replace_parent(&mut self, x: &NodeRef<T, V>, y: &NodeRef<T, V>) {
+        y.set_parent(x.parent_owned());
+        if x.parent(NodeRef::is_sentinel) {
+            self.root = y.clone_rc();
+        } else if x.parent(|p| p.left(|l| l.eq(x))) {
```

Comment:
Please use `is_left_child` to replace this condiction.

Reply:
done